homeassistant/components/hyperion/config_flow.py

```diff
@@ -0,0 +1,380 @@
+        # Sample data provided by SSDP: {
+        #   'ssdp_location': 'http://192.168.0.1:8090/description.xml',
+        #   'ssdp_st': 'upnp:rootdevice',
+        #   'deviceType': 'urn:schemas-upnp-org:device:Basic:1',
+        #   'friendlyName': 'Hyperion (192.168.0.1)',
+        #   'manufacturer': 'Hyperion Open Source Ambient Lighting',
+        #   'manufacturerURL': 'https://www.hyperion-project.org',
+        #   'modelDescription': 'Hyperion Open Source Ambient Light',
+        #   'modelName': 'Hyperion',
+        #   'modelNumber': '2.0.0-alpha.8',
+        #   'modelURL': 'https://www.hyperion-project.org',
+        #   'serialNumber': 'f9aab089-f85a-55cf-b7c1-222a72faebe9',
+        #   'UDN': 'uuid:f9aab089-f85a-55cf-b7c1-222a72faebe9',
+        #   'ports': {
+        #       'jsonServer': '19444',
+        #       'sslServer': '8092',
+        #       'protoBuffer': '19445',
+        #       'flatBuffer': '19400'
+        #   },
+        #   'presentationURL': 'index.html',
+        #   'iconList': {
+        #       'icon': {
+        #           'mimetype': 'image/png',
+        #           'height': '100',
+        #           'width': '100',
+        #           'depth': '32',
+        #           'url': 'img/hyperion/ssdp_icon.png'
+        #       }
+        #   },
+        #   'ssdp_usn': 'uuid:f9aab089-f85a-55cf-b7c1-222a72faebe9',
+        #   'ssdp_ext': '',
+        #   'ssdp_server': 'Raspbian GNU/Linux 10 (buster)/10 UPnP/1.0 Hyperion/2.0.0-alpha.8'}
+
+        # SSDP requires user confirmation.
+        self._require_confirm = True
+
+        self._data = {}
+        self._data[CONF_HOST] = urlparse(discovery_info[ATTR_SSDP_LOCATION]).hostname
+        try:
+            self._port_ui = urlparse(discovery_info[ATTR_SSDP_LOCATION]).port
+        except ValueError:
+            self._port_ui = const.DEFAULT_PORT_UI
+
+        try:
+            self._data[CONF_PORT] = int(
+                discovery_info.get("ports", {}).get(
+                    "jsonServer", const.DEFAULT_PORT_JSON
+                )
+            )
+        except ValueError:
+            self._data[CONF_PORT] = const.DEFAULT_PORT_JSON
+
+        hyperion_id = discovery_info.get(ATTR_UPNP_SERIAL)
+        if hyperion_id:
+            # For discovery mechanisms, we set the unique_id as early as possible to
+            # avoid discovery popping up a duplicate on the screen. The unique_id is set
+            # authoritatively later in the flow by asking the server to confirm its id
+            # (which should theoretically be the same as specified here)
+            await self.async_set_unique_id(hyperion_id)
+            self._abort_if_unique_id_configured()
+        else:
+            return self.async_abort(reason="no_id")
+        return await self.async_step_auth()
+
+    # pylint: disable=arguments-differ
+    async def async_step_user(
+        self,
+        user_input: Optional[ConfigType] = None,
+        errors: Optional[Dict] = None,
```

Comment:
The step functions should not accept other parameters than the user_input and the form errors are private to each step since they are specific to the form.

Reply:
@MartinHjelmare -- help me out here, as it feels like these constraints:

   * An inability to pass in errors into a step.
   * An inability to return another step's form.

... make it impossible (?) to show an error message to a user in the context of  a different step.

An example:
   * The `user` steps get host/port (may or may not be verified at the time).
   * A future step uses that host/port, but fails to connect.

Desired outcome:
   * We'd want  to return the user to the host/port choice, with an error message saying `Can't connect`.

How can I achieve this, or do you have an alternative suggestion?

Reply:
If the user step form fails there should not be a different step after the user step. Each step should be validated before moving on to the next step. We should never need to return to a previous step.

Reply:
OK. What behavior would you like if the subsequent connection fails in a future step, even though it was verified in a prior step? Abort?

Reply:
Yeah, since that doesn't have to do with the user input.